Das/pytutorial 2 (#52)
Arcadia-Science/arcadia-computational-training: docs/arcadia-users-group/20230314-intro-to-python-2/lesson.md

# Introduction to Python, Part 2

Follow along in **Binder** by clicking the badge below:  

> [![Binder](https://mybinder.org/badge_logo.svg)](https://mybinder.org/v2/gh/Arcadia-Science/arcadia-computational-training/das/pytutorial?labpath=docs%2Farcadia-users-group%2F20230314-intro-to-python-2%2Flesson.ipynb)

Reviewer: can you update this now to main branch so it will resolve from there? i know it will break rn if we click it but then it's the one we want to use for the actual lesson

Author: Updated!

---
## 0. Loops and Conditionals

This lesson is the second in a series of workshops that provide an introduction to programming using Python.

**This lesson** builds on the foundation of lesson 1 and covers:
- lists and dictionaries
- methods
- loops
- booleans
- conditional expressions

---
## 1. Data structures

Often, when you're working with values in science, you're not just working with `x = 1` and `y = 2`, but instead with large amounts of data. Python offers a variety of data structures that aggregate data, including:

- `list`s: ordered arrays of values.
- `dict` or dictionaries: values ordered with key-value pairs.
- `set`s: like lists, but each value can only occur once.
- `tuple`s: like lists, but you can't modify the values.

> We'll first talk about `list` and `dict` structures, as they're the easiest and most intuitive to understand.

Reviewer: ```suggestion
We'll first talk about `list` and `dict` structures, as they're the most common data structures.
```

---
### 1.1 Lists
You can create a `list` in Python using bracket notation, such as below.

```python
lst = [1, 2, 3, 4, 5]
```

To access a specific value in the list, such as `2`, you use the **index**, or the position of the value in the list, in bracket notation, as below.

> A key thing to remember in Python is that **lists are indexed starting with the index `[0]`**. So, if you wanted to get the value `2` from the list above, you would use an the index `[1]`.

Reviewer: ```suggestion
A key thing to remember in Python is that **lists are indexed starting with the index `[0]`**. So, if you wanted to get the value `2` from the list above, you would use the index `[1]`.
```

```python
lst[1]
```

    2

Lists are powerful because you can do **many** different operations with them using built-in functionality. For example, you can slice lists to get only specific elements using the colon operator `:`.

To get the first through third elements of a list, you would use the following notation:

```python
lst[0:3]
```

    [1, 2, 3]

> Note that, in the above example, the elements are extracted with a 'left-included, right-excluded' notation - the *fourth* element in the list, `4`, at index `3`, is not included.

Reviewer: nice, glad you included this

You can slice and index lists in some pretty interesting ways, such as in the examples below:

```python
lst = [0, 1, 2, 3, 4, 5, 6, 7, 8]

# The below expression is equivalent to lst[0:3].
# It gives you everything from the start of the list up to the ending address.
print(lst[:3])

# The below expression gives you the last element in the list.
print(lst[-1])

# The below expression gives you the last two elements in the list.
print(lst[-2:])
```

    [0, 1, 2]
    8
    [7, 8]

> Note that accessing a slice of a `list` returns a list, whereas accessing a single element returns the value itself.

Reviewer: good note!

This method allows you to replace a bracketed space or variable name into a string, as below.

```python
example_1 = "The quick {} fox jumps over the {} dog.".format("red", "happy")
example_2 = "The quick {0} fox jumps over the {1} dog.".format("arctic", "sleepy")
example_3 = "The quick {foxtype} fox jumps over the {dogtype} dog.".format(foxtype = "silver", dogtype = "lucky")

print(example_1, example_2, example_3, sep = '\n')
```

    The quick red fox jumps over the happy dog.
    The quick arctic fox jumps over the sleepy dog.
    The quick silver fox jumps over the lucky dog.

> ### Practice 2
> a) How can you use `.format()` to adjust the following text so that it displays the price in dollar notation (\$1.00)?"  
> b) How could you convert this to instead display things in yen format (¥100) where one dollar = 100 yen?

<details>
    <summary> Practice 2 Answer </summary><br>
    a) <code>print("Apples cost ${price:.2f} each.".format(price = cost))</code><br>
    b) <code>print("Apples cost ¥{price} each.".format(price = cost * 100))</code><br>
</details>

```python
cost = 1

print("Apples cost {price} each.".format(price = cost))
```

>     Apples cost 1 each.

Reviewer: lol 1 each

Prior to the loop, we specify a variable `i`. At the end of each entry in the loop, we increment `i` by 1 in order to step through the list `a`.

How do we know when to stop? Take a look at the **`while`** statement:
```python
while a[i] < 3:
```
This means that as long as the entry at index `[i]` in list `[a]` is less than 3, we continue through the loop, printing the value of `a[i]`.

As soon as we arrive at `i = 3`, at the top of the **`while`** loop, we get the value `a[i] = 3`, which is NOT less than `3`. Therefore, the **`while`** loop terminates, and the value `3` is never printed.

**`while`** loops can be very powerful for a class of Python functions that rely on recursion, but you can also easily end up writing functions that "hang", or run forever without stopping.

## P2. Practice

Let's explore using loops.

> ### Practice 3
> Write a function called `fasta_renamer()` which takes a list of filenames and renames them with the following specifications:
> - Converts all characters to lowercase.
> - Converts files ending with `.fasta` to `.fa`.
> - Optionally, adds a prefix of the user's choice to the start of every file name.

> > - Returns a list of the new file names.

Reviewer: this might be a slightly confusing example because the files aren't written anywhere on the system, and they aren't derived from running `ls` or something on the system. i think clarifying that might be enough, or you could switch the example to sample names again, or a list of genbank or uniprot identifiers or something

Author: That makes sense - I updated this to mention that this just a list of file name strings, not actual files, and that we'll be learning how to open/read files in the next lesson.

```python
names = ['Celine', 'Britney', 'Stefani', 'Rina', 'Charli', 'Billie']
output = []

for name in names:
    if name[0] == 'C':
        output.append(name + '_01')
    else:
        continue

print(output)
```

    ['Celine_01', 'Charli_01']

### 4.4 Other conditional statements

There are other types of conditional statements in Python which you'll encounter less frequently, but can still be very helpful:

- **`try`** **`except`** **`finally`**: tries to do something, unless it returns an error.  
    If an error occurs, runs the **`except`** statement. You can specify different outcomes for different error types.  
    **`finally`** at the end of each attempt, do a thing.

```python
values = ['a', 1, 3, '5', 'q', 3.0]

for value in values:
    try:
        print(value / 3)
    except TypeError:
        print("Cannot divide - wrong value type.")
    finally:
        print("---")
```

    Cannot divide - wrong value type.
    ---
    0.3333333333333333
    ---
    1.0
    ---
    Cannot divide - wrong value type.
    ---
    Cannot divide - wrong value type.
    ---
    1.0
    ---

> - **`match`** **`case`**: works very similarly to an if - elif - else statement, but can be much easier to read when you have specific expected outcomes. See example [here](https://learnpython.com/blog/python-match-case-statement/).

Reviewer: ```suggestion
- **`match`** **`case`**: works very similarly to an `if` - `elif` - `else` statement, but can be much easier to read when you have specific expected outcomes. See example [here](https://learnpython.com/blog/python-match-case-statement/).
```
